src/samplers/usercall/mod.rs

```diff
@@ -0,0 +1,283 @@
+// Copyright 2019-2021 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use async_trait::async_trait;
+
+use std::sync::{Arc, Mutex};
+use std::time::Instant;
+
+use crate::common::bpf::BPF;
+use crate::config::SamplerConfig;
+use crate::samplers::{Common, Sampler};
+
+#[cfg(feature = "bpf")]
+use crate::common::bpf::bpf_hash_char_to_map;
+
+#[cfg(feature = "bpf")]
+use rustcommon_metrics::Statistic;
+
+mod config;
+mod stat;
+
+pub use config::{LibFileMap, LibSearchMap, UsercallConfig};
+pub use stat::{UsercallStatistic, NAMESPACE};
+
+use std::path::Path;
+
+#[allow(dead_code)]
+pub struct Usercall {
+    bpf: Option<Arc<Mutex<BPF>>>,
+    bpf_last: Arc<Mutex<Instant>>,
+    common: Common,
+    statistics: Vec<UsercallStatistic>,
+    lib_files: LibFileMap,
+    lib_search: LibSearchMap,
+}
+
+const DEFAULT_LIB_SEARCH_PATHS: [&str; 6] = [
+    "/lib64",
+    "/usr/lib64",
+    "/usr/local/lib64",
+    "/lib",
+    "/usr/lib",
+    "/usr/local/lib",
+];
+
+#[cfg(feature = "bpf")]
+const PROBE_PRELUDE: &str = r#"
+#include <uapi/linux/ptrace.h>
+struct key_t {
+    char c[80];
+};
+BPF_HASH(counts, struct key_t);
+
+"#;
+
+macro_rules! probe_template {
+    () => {
+        r#"
+
+int probe_{}(void *ctx) {{
+    struct key_t key = {{.c = "{}"}};
+    u64 zero = 0, *val;
+    val = counts.lookup_or_init(&key, &zero);
+    (*val)++;
+    return 0;
+}}
+"#
+    };
+}
+
+#[allow(dead_code)]
+fn path_match(lib_name: &str, path: &Path) -> bool {
```

Comment:
Might be worth adding a comment along the lines of your response to my asking about why this is string manipulation. It's not immediately obvious from the code, so worth putting a hint here as to why we need to do it this way.